### ISSUE TITLE
bilinear_filter and trilinear_filter are a disaster

### ISSUE THREAD

**Issue by sfan5** [luanti-org/luanti#15604]
Luanti f2b1cc3e6 on Linux
Renderer: 4.6 / opengl3 / SDL

so this is the average experience if you enable one or both of bilinear/trilinear filtering:
![screenshot_20241229_201312](https://github.com/user-attachments/assets/3d2a5958-9dee-41d7-8380-0be5610a404b)

how it *should* look like:
![screenshot_20241229_201536](https://github.com/user-attachments/assets/005d2cbf-9cf5-480e-8b9e-0f4de36fecf1)

now internally the code already upscales textures to 64px in an attempt to make them look less awful (as described [here](https://github.com/minetest/minetest/blob/f2b1cc3e6141f43f8e7b5e55f86ccc05f7b3b1bf/builtin/settingtypes.txt#L1884-L1892)), but this is not nearly enough.
someone had the foresight to solve this differently for CAOs and the filtering is disabled for textures smaller than 65px or they'd look like a blurry mess too.
wield meshes have yet another hacky cutoff at 33px ([code](https://github.com/minetest/minetest/blob/294a30e445ec9cd94d66efc8f94048606716ad35/src/client/wieldmesh.cpp#L285-L291))

let's recap the problems:
* graphics turn to shit if used with the pixel-art style that is very common in Luanti
* inconsistently applied throughout the engine
* even if not, modders have no say in this

since the settings redesign these are well-enough hidden that people probably don't usually enable them (good), but who is actually using it *intentionally*? in which use case?
I propose removing it until we've come up with a better way.

**Comment by sfan5** [comment]
well I guess there is nothing to discuss as @Desour has proposed this before: #13108

**Comment by nininik0** [comment]
This also sometimes lags especially on heavy servers.

**Comment by grorp** [comment]
> who is actually using it intentionally? in which use case?

people playing with high res texture packs (there are a few popular ones on ContentDB) and people playing games with high res textures (there is Extra Ordinance) probably

(these are obviously cases where content creators instead of users should get control)

> (note: this applies to `anisotropic_filter` too)

as dicussed previously, this apparently depends on your GPU. on my hardware, anisotropic filtering works fine and doesn't result in blurriness / bilinear filtering.

**Comment by sfan5** [comment]
> people playing with high res texture packs probably

That makes sense and I forsee these people being upset if we take this feature away.
A possible compromise would be to consistently apply a minimum texture size below which the filtering isn't enabled. For example 256px.

> this apparently depends on your GPU. on my hardware, anisotropic filtering works fine and doesn't result in blurriness / bilinear filtering.

are you sure?
on its own aniso doesn't do anything, because it just modifies filter behavior. together with bilinear it produces the same mess:
![screenshot_20241230_163322](https://github.com/user-attachments/assets/b56cebda-cecb-482d-91ae-7de977f561a3)

**Comment by grorp** [comment]
> are you sure?
> on its own aniso doesn't do anything, because it just modifies filter behavior. together with bilinear it produces the same mess: [image]

yes, anisotropic filtering does have an effect on its own (on my setup):

<p>all settings at their default values<br><img src="https://github.com/user-attachments/assets/4fa63bb0-7ea7-4d33-9539-9eb3bf941c07" alt="" width="512"></p>

<p>aniso enabled, all other settings at their default values<br><img src="https://github.com/user-attachments/assets/69ee97c3-72e1-47fc-a7da-d15e59156c19" alt=""  width="512"></p>

it improves the look of the brick wall and far-away snow while introducing no blurriness problems. enabling aniso also gets rid of artifacts caused by mipmapping (on my setup):

<p>mipmapping enabled, all other settings at their default values<br><img src="https://github.com/user-attachments/assets/af3f58e1-8eb6-4edb-90dc-c390eb6f6d77" alt="" width="512"></p>

<p>mipmapping + aniso enabled, all other settings at their default values<br><img src="https://github.com/user-attachments/assets/e969eedb-c756-4145-a1c7-d00295bcd977" alt="" width="512"><br>(this looks <em>very</em> similar to the screenshot with aniso only, I've confirmed this two times to be sure since it was unexpected)</p>

NVIDIA GeForce GTX 1080 Ti, 4.6.0 / opengl3 / SDL

**Comment by sfan5** [comment]
> yes, anisotropic filtering does have an effect on its own (on my setup):

Hmm, I'll do some more testing then.

**Comment by sfan5** [comment]
> that sounds reasonable. I'd choose a lower threshold of 128px or 64px.

disagree. everything is noticably blurry (and that is without zoom or standing unreasonably close) at 64 or 128.
192 is the first value I'd give to unsuspecting players in good conscience.

**Comment by sfan5** [comment]
```
21:28 <+sfan5> my takeaway is that texture_min_size is the magic moire-removal knob for some people so we can't get rid of it just yet
21:28 <+sfan5> also means the only way to achieve my goal is if we set a high minimum on texture_min_size to prevent people from turning their game into blur 
```

**Comment by lhofhansl** [comment]
> someone had the foresight to solve this differently for CAOs and the filtering is disabled for textures smaller than 65px or they'd look like a blurry mess too

#6562? Yes, that was annoying, and kinda a quick hack.

**Comment by appgurueu** [comment]
> Anyone with interest over visual direction in their game would consider it insane to provide a 'blurrymator 9000' knob in graphics settings, for the user to decide.

I agree. glTF lets you choose filtering settings. How should that interact with such settings? There's sort of an ugly compromise where you use the player settings as a default and give the modder control otherwise, but it would imo be best if this was solely the game developers' decision.

**Comment by sfan5** [comment]
> How should that interact with such settings?

In the long run we will probably remove these settings. I'd ideally like to do that immediately but as it now it breaks someone's [workflow](https://xkcd.com/1172/) so we can't be as hasty.

**Comment by sfan5** [comment]
My PR now proposes a consistent and documented interim solution, without breaking the combination of settings @Warr1024 described on IRC.
